python/ray/data/_internal/execution/operators/hash_shuffle.py

```diff
@@ -445,12 +446,152 @@ def __init__(
+                    f"Required: {required_memory / GiB:.2f} GiB for {self._aggregator_pool.num_aggregators} aggregators, "
+                    f"but cluster only has {total_memory / GiB:.2f} GiB total memory. "
+                    f"Consider reducing the number of partitions or increasing cluster size."
+                )
+
+            if required_memory > available_memory:
+                logger.warning(
+                    f"Limited available memory resources for hash shuffle operation. "
+                    f"Required: {required_memory / GiB:.2f} GiB, available: {available_memory / GiB:.2f} GiB. "
+                    f"Aggregators may take longer to start due to resource contention."
+                )
+
+        logger.debug(
+            f"Resource check passed for hash shuffle operation: "
+            f"required CPUs={required_cpus}, available CPUs={available_cpus}, "
+            f"required memory={required_memory / GiB:.2f} GiB, available memory={available_memory / GiB:.2f} GiB"
+        )
+
+    def _check_aggregator_health(self) -> None:
+        """Check if all aggregators are up and running after a timeout period.
+        Uses non-blocking ray.wait to check actor readiness.
+        Will warn every 10 seconds if aggregators remain unhealthy.
+        """
+        min_wait_time = self.data_context.min_hash_shuffle_aggregator_wait_time_in_secs
+        if (
+            self._aggregator_startup_time is None
+            or time.time() - self._aggregator_startup_time < min_wait_time
+        ):
+            return
+
+        try:
+            # Initialize readiness refs if not done yet
+            if self._readiness_refs is None:
+                self._readiness_refs = [
+                    aggregator.__ray_ready__.remote()
+                    for aggregator in self._aggregator_pool._aggregators
+                ]
+
+            # Use ray.wait to check readiness in non-blocking fashion
+            _, unready_refs = ray.wait(
+                self._readiness_refs,
+                num_returns=len(self._readiness_refs),
+                timeout=0.1,  # Short timeout to avoid blocking
+            )
+
+            current_time = time.time()
+            should_warn = unready_refs and (  # If any refs are not ready
+                self._last_health_warning_time is None
+                or current_time - self._last_health_warning_time
+                >= self._health_warning_interval
+            )
+
+            if should_warn:
+                logger.warning(
+                    f"Hash shuffle aggregator health check failed after {min_wait_time}s. "
+                    f"{len(unready_refs)}/{self._aggregator_pool.num_aggregators} "
+                    f"This may indicate resource constraints or cluster issues. "
+                    f"The operation may proceed slowly or fail. "
+                    f"Will continue checking every {self._health_warning_interval}s."
```

Comment:
Let's shape message along the lines of:

"Only X out of Y hash-shuffle Aggregators are ready after K min. This might indicate resource contention for cluster resources (available CPUs: N, required CPUs: M), consider increasing cluster size."